File: old/recDP/002-fib2.py
def fibo(n):
    fib = {0: 0, 1: 1}
    for i in range(2, n + 1): # Bug start from 2
        fib[i] = fib[i - 1] + fib[i - 2]
    return fib[n]

print(fibo(6)) # 8
print(fibo(4)) # 3

# overlapping (sub)problems
# optimal (sub)structure

#           fib(8)
#   fib(7)         fib(6)
# fib(6) fib(5)    .  .

# first write the general case i.e. the recurrence breakdown: assume you'll get the lower answers and asssume you have given the base cases
#   then, write all the base cases
#     then, execute(execution context,etc) and check
#       [then, the O(T(n)), O(S(n))]

File: old/recDP/002-fib2.cpp
#include <bits/stdc++.h>
using namespace std;


int fib(int n) {
    vector<int> dp(n + 1, 0);
    dp[0] = 0;
    dp[1] = 1;
    for (int i = 2; i <= n; i++) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }

    return dp[n];
}

int main(void) {
    cout << fib(5) << endl;
    cout << fib(6) << endl;
    cout << fib(8) << endl;
}
